// ngx-image-gallery/esm5/lib/index.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxImageGalleryComponent } from './components/ngx-image-gallery/ngx-image-gallery.component';
import { ClickOutsideDirective } from './directives/click-outside.directive';
import { MouseWheelDirective } from './directives/mousewheel.directive';
import * as ɵngcc0 from '@angular/core';
export * from './components/ngx-image-gallery/ngx-image-gallery.component';
export * from './directives/click-outside.directive';
export * from './directives/mousewheel.directive';
var NgxImageGalleryModule = /** @class */ (function () {
    function NgxImageGalleryModule() {
    }
NgxImageGalleryModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxImageGalleryModule });
NgxImageGalleryModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxImageGalleryModule_Factory(t) { return new (t || NgxImageGalleryModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxImageGalleryModule, { declarations: function () { return [NgxImageGalleryComponent,
        MouseWheelDirective,
        ClickOutsideDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxImageGalleryComponent,
        MouseWheelDirective,
        ClickOutsideDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxImageGalleryModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    NgxImageGalleryComponent,
                    MouseWheelDirective,
                    ClickOutsideDirective
                ],
                exports: [
                    NgxImageGalleryComponent,
                    MouseWheelDirective,
                    ClickOutsideDirective
                ]
            }]
    }], function () { return []; }, null); })();
    return NgxImageGalleryModule;
}());
export { NgxImageGalleryModule };
